### DISCUSSION TITLE
Annotation to copy from Source PDF page to Target PDF page..

### DISCUSSION THREAD

**Discussion by pachhaipurna** [pymupdf/PyMuPDF#2858] [Looking for help]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Potentially add an issue reference.

**Describe the solution you'd like**

**Describe alternatives you've considered**
Are there several options for how your request could be met?

**Additional context**

**Comment by pachhaipurna** [comment]
Hello @JorjMcKie ,

I am currently trying to copy annotation form source pdf page to Target pdf page,
I have copied some annotation based on Type, However still unable to copy all annotation to target page, Like Trianlge type Stamp other annotation.
Is there any way we can copy all annotation object at once or, Just to copy annotation, Could you please give me some idea ?

Here is my code.
`
import fitz



"""Copy annotations from Source PDF to Target PDF"""
def copy_annotations(source_path,target_path):

    ##Source PDF processing
    source_pdf = fitz.open(source_path,filetype = "pdf")
    source_page = source_pdf[1]

    ##Target PDF processing..
    target_pdf = fitz.open(target_path,filetype = "pdf")
    target_page = target_pdf[0]

    ###Read Input PDF page based on Page index. page_values[0] is the index
    if source_pdf.has_annots():
        # target_page.xref_copy(source_page)
        for source_annotation in source_page.annots():
            
            for annot_type in source_annotation.type:

                if annot_type =='Circle':
                    target_annotation = target_page.add_circle_annot(source_annotation.rect)
                    text_rect = fitz.Rect(source_annotation.rect.x0,source_annotation.rect.y0,source_annotation.rect.x1,source_annotation.rect.y1)
                    target_annotation.set_colors(source_annotation.colors)

                    target_annotation.set_border(source_annotation.border)
                    target_annotation.set_opacity(source_annotation.opacity)
                    target_annotation.set_info(subject = source_annotation.info['subject'])
                    target_annotation.update()
                    logger.info('Circle Annotation Added..')

       
                elif annot_type == 'FreeText':
                 
                    target_annotation = target_page.add_freetext_annot(source_annotation.rect,source_annotation.info['content'])
                    
                    target_annotation.set_colors(source_annotation.colors)
                    target_annotation.set_border(source_annotation.border)
                    target_annotation.set_opacity(source_annotation.opacity)
                    logger.info('FreeText Annotation Added..')
                    target_annotation.set_info(subject = source_annotation.info['subject'])
                    target_annotation.update()


                elif annot_type == 'Stamp':
                    
                    print('Xref', source_annotation.quad)
                    target_annotation = target_page.add_stamp_annot(source_annotation.rect, stamp =0)
                    target_annotation.set_colors(source_annotation.colors)
                    target_annotation.set_border(source_annotation.border)
                    target_annotation.set_opacity(source_annotation.opacity)

                    target_annotation.set_info(subject = source_annotation.info['subject'])
                    logger.info('Stamp Annotation Added..')
                    target_annotation.update()

                

                ###If annotation type is Line 
                elif annot_type == 'Line':

                    line_coords = source_annotation.vertices
                    target_annotation = target_page.add_line_annot(line_coords[0], line_coords[1])
                    target_annotation.set_colors(source_annotation.colors)
                    target_annotation.set_border(source_annotation.border)
                    target_annotation.set_opacity(source_annotation.opacity)
                    target_annotation.set_info(subject = source_annotation.info['subject'])
                    target_annotation.update()
                    logger.info('Line Annotation Added..')


                elif annot_type == 'PolyLine':
                    target_annotation = target_page.add_polyline_annot(source_annotation.vertices)
                    target_annotation.set_colors(source_annotation.colors)
                    target_annotation.set_border(source_annotation.border)
                    target_annotation.set_opacity(source_annotation.opacity)
                    target_annotation.set_info(subject = source_annotation.info['subject'])
                    logger.info('PolyLine Annotation Added..')
                    target_annotation.update()

            
                elif annot_type == 'Polygon':

                    target_annotation = target_page.add_polygon_annot(source_annotation.vertices)
                    target_annotation.set_colors(source_annotation.colors)
                    target_annotation.set_border(source_annotation.border)
                    target_annotation.set_opacity(source_annotation.opacity)

                    # target_annotation.set_line_ends(source_annotation.line_ends)
                    target_annotation.set_info(subject = source_annotation.info['subject'])
                    target_annotation.update()
                   

                print(source_page.get_xobjects(),'xref bassed')

    else:
        logger.info('No annotation')

    target_pdf.save(target_path,encryption=0)
    target_pdf.close()
    source_pdf.close()


if __name__ == '__main__':

    source_path ='source.pdf'
    target_path = 'target.pdf'
    copy_annotations(source_path,target_path)`

![image](https://github.com/pymupdf/PyMuPDF/assets/43229248/b2aad358-2db9-4869-a65c-36b2809bcccd)

These types of annotation are not copied .Actually all those annotation are prepared form Bluebeam software.
Please help.

**Comment by pachhaipurna** [comment]
@JorjMcKie If there are other options Could you please let me know. I am not an expert in Python so need help.

I need to copy whole annotation to target pdf page, The target page will have no any annotation at first so need to update the target page with source page annotation.

**Comment by JorjMcKie** [comment]
I have moved your post to Discussions - it is no _issue_.

The not working part (triangle etc.) are no annotations - at least no standard annotations, so there is no way to re-generate them on the target.
In PDF there exists no object like a triangle. The PDF creator has created this with own proprietary code. Same is true for those cloudy objects. If those are annotations at all, then the creator must have abused an existing annotation type and generated the appearance by own code.

This could only be replicated by low-level code.

You are aware that you can simply copy the source page to the target PDF as a page there? Use `doc.insert_pdf(source, from_page=n, to_page=n, ...)`.
This will copy the source page, with all annotations.

**Reply by pachhaipurna** [reply]
@JorjMcKie Except annotations pdf are different like drawing in pdf 's are different so Just need to copy those annotation.
Or Is there any option to replace the contents of pdf's except annotation.

**Reply by pachhaipurna** [reply]
Actually some portion of target pdf are updated one, SO Just need to copy annotation to target page, Like Some text contents and Date or some diagrams are added in target page,

**Comment by JorjMcKie** [comment]
> @JorjMcKie Except annotations pdf are different like drawing in pdf 's are different so Just need to copy those annotation. Or Is there any option to replace the contents of pdf's except annotation.

Annotations in general are no part of the page's contents - they are like "above" of the rest. If I understand you correctly, you could take the source page (with those funky annotations) and overlay it with the target page's content using `show_pdf_page()`.
The source page in this case should still show its annotations after that.

**Reply by pachhaipurna** [reply]
@JorjMcKie Thanks it is working now, But I have one issue, For "FreeTextCallout" annotations are removed once the add_redact_annot() is used, if I don't use the contents are overlapped.
Any suggestions on this

`source_page.add_redact_annot(target_page.rect,fill=(1,1,1))
    source_page.apply_redactions()

    # page_front.insert_pdf(target_pdf, from_page=0, to_page=0)
    source_page.show_pdf_page(target_page.rect, target_pdf, pno=0, keep_proportion=True, overlay=True, oc=0, rotate=0, clip=None)

    source_pdf.save(updated_path, encryption=fitz.PDF_ENCRYPT_KEEP,garbage=4,
         clean=True,
         deflate=True,
         deflate_images=True,
         deflate_fonts=True)`

![image](https://github.com/pymupdf/PyMuPDF/assets/43229248/f17d97f4-7f25-4f79-a969-34dd4397d114)

These callouts are removed once I used add_redact_annot(), If I dont use that the letters are overlapped .

**Reply by pachhaipurna** [reply]
These FreeTextCallout are present in source page annots. No need to replace these annots except pdf contents.

I don't want to remove those type of annotations, Please help

**Comment by pachhaipurna** [comment]
@JorjMcKie  It supports Callout or not, I have one issue When I use add_redact_annot() before using show_pdf_page() all Text annotations as well as FreetextCalout annotations are also redacted, Any idea not to remove any Freetext and annotations.

**Reply by JorjMcKie** [reply]
🤷‍♂️sorry

**Reply by pachhaipurna** [reply]
@JorjMcKie I am able to replace pdf page without removing the target page annotations using "draw_rect()".
However some of the pdf page orientation is not same and I rotated the page using set_rotation().

There are few issues some page seems landscape but the coordinates are totally different and while replacing the target page source [age is replacing in specific area only, Is there any option to replace the whole page with same layout orientation like need to check the all 4 corner coordinates.

Could you please let me help me ?
Thank you.

I have another query, Is it possible to update the annotation content value "Generated using Bluebeam", I tried to update, but it is removing the colors and annotation attribute also, Any solution on this ?

![image](https://github.com/pymupdf/PyMuPDF/assets/43229248/16e67d4e-420b-4b47-bf90-2aeac57712b3)
Original annotation content is this..

after update.

![image](https://github.com/pymupdf/PyMuPDF/assets/43229248/1788907b-86d8-4ca3-b791-a88c8de446c9)


here the border and arrow of annotation are removed, Any option to control these attributes.

**Reply by JorjMcKie** [reply]
Like with any other package, PyMuPDF supports is own subset of annotation features. If coming across an annotation created by a different package, re-creating that annotation with any other package than the original one is highly likely to fail.
The same is true for any other software: you will see limitations in terms of fonts, border line dashing, fill colors, line-end formats etc. if re-creating an annotation with package B that has been made by package A.

PyMuPDF does not support callout arrows with FreeText annotations.
